### ISSUE TITLE
Caddy environment files do not affect config & data paths

### ISSUE THREAD

**Issue by pascal-zarrad** [caddyserver/caddy#4273]
### 1a. Operating system and version

```
1. Ubuntu 20.04
2. Alpine 3.14.1
```

### 1b. Caddy version (run `caddy version` or paste commit SHA)

```
1. v2.4.3 h1:Y1FaV2N4WO3rBqxSYA8UZsZTQdN+PwcoOcAiZTM8C0I=
2. 569ecdbd02a3bdf8589f785cca022240973dea4d
```

### 1c. Go version

```
1.15
```

### 2a. What happens
I start go with the `--envfile` flag and point it to a envfile with the following contents:
```
XDG_DATA_HOME=/home/[user]/.caddy/data
XDG_CONFIG_HOME=/home/[user]/.caddy/config
```

I start Caddy now with `caddy run --envfile /home/[user]/.caddy/.env --config /home/[user]/Caddyfile --environ`.

I would now expect Caddy to put its files into the specified directories. Instead, Caddy will create the directories in the default location.

When using `--environ`, the log output of the environment variables shows the proper values (except for `caddy.ConfigAutosavePath`).
However, in the further output where Caddy informs about the autosaved config, the common paths are used. 

Using real environment variables works as expected and the files are saved at the proper location.

### 2b. Why it's a bug
It is expected that the .env file has the same effect as setting the environment variable on the system.

### 2c. Log output

Without a fix (Stripped down to the essential parts):
```
caddy.HomeDir=/home/pascal/caddydir/ch
caddy.AppDataDir=/home/pascal/caddydir/cd/caddy
caddy.AppConfigDir=/home/pascal/caddydir/cc/caddy
caddy.ConfigAutosavePath=/home/pascal/.config/caddy/autosave.json
caddy.Version=(devel)

...
{"level":"info","ts":1628361986.0723467,"msg":"using provided configuration","config_file":"/home/pascal/caddy.json","config_adapter":""}
2021/08/07 20:46:26 [INFO] Redirecting sink to: stdout
{"level":"warn","ts":1628361986.0747495,"logger":"admin","msg":"admin endpoint disabled"}
{"level":"info","ts":1628361986.0792449,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc000508000"}
{"level":"info","ts":1628361986.0966988,"msg":"autosaved config (load with --resume flag)","file":"/home/pascal/.config/caddy/autosave.json"}
{"level":"info","ts":1628361986.096718,"msg":"serving initial configuration"}
{"level":"info","ts":1628361986.096729,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/home/pascal/.local/share/caddy"}
{"level":"info","ts":1628361986.0967355,"logger":"tls","msg":"finished cleaning storage units"}
```

With a fix (Stripped down to the essential parts):
```
caddy.HomeDir=/home/pascal/caddydir/ch
caddy.AppDataDir=/home/pascal/caddydir/cd/caddy
caddy.AppConfigDir=/home/pascal/caddydir/cc/caddy
caddy.ConfigAutosavePath=/home/pascal/caddydir/cc/caddy/autosave.json
...
{"level":"info","ts":1628362230.0860615,"msg":"using provided configuration","config_file":"/home/pascal/caddy.json","config_adapter":""}
2021/08/07 20:50:30 [INFO] Redirecting sink to: stdout
{"level":"warn","ts":1628362230.0936966,"logger":"admin","msg":"admin endpoint disabled"}
{"level":"info","ts":1628362230.0943089,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc000230070"}
{"level":"info","ts":1628362230.112259,"msg":"autosaved config (load with --resume flag)","file":"/home/pascal/caddydir/cc/caddy/autosave.json"}
{"level":"info","ts":1628362230.1122801,"msg":"serving initial configuration"}
{"level":"info","ts":1628362230.113837,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/home/pascal/caddydir/cd/caddy"}
{"level":"info","ts":1628362230.113868,"logger":"tls","msg":"finished cleaning storage units"}
```

### 2d. Workaround(s)

1. Use system environment variables

Only other work around is to really solve the issue.

## 3. Tutorial (minimal steps to reproduce the bug)

1. Add a .env file (as described below
2. Add a basic Caddyfile (can contain anything, you just need any)
3. Start Caddy using `caddy run --envfile <path-to-env-file> --config <path-to-config>`
4. Check if Caddy created its files at the location specified in the env file

**.env file:**
```
XDG_DATA_HOME=/home/[user]/.caddy/data
XDG_CONFIG_HOME=/home/[user]/.caddy/config
```

**Caddyfile:**
```
localhost {
        root: /path/to/any/route
	encode zstd gzip
	templates
	file_server browse
}
``` 

You can try the same test again using system environment variables (e. g. `export XDG_DATA_HOME=/home/[user]/.caddy/data`) and you will notice that it worked as expected.

## 4. Solution
The issue occurs because the values are being initialized as soon as `storage.go` is used (`var xy = xy.xy`).
A possible solution would be to add a reinitialization/update of `caddy.ConfigAutosavePath` and `caddy.DefaultStorage`.

This could be done by adding a call to some function (e. g. `UpdateStoragePaths `or `ReInitializeStoragePaths` after updating the environment variables in `cmd/main.go#loadEnvFromFile` after line 362.

Then we could add a function to `storage.go` that just runs the code used to initialize the variables again. This would only being called when really a env file has been specified and not at all affect anything when not using one.

Example:
```go
// file: cmd/main.go
// line: 364
caddy.UpdateStoragePaths()
```

```go
// file: storage.go
// starting at line: 156
// --> This is the newly added function
// UpdateStoragePaths sets the config auto save and default storage path
// again. This ensures that env files that overwrite the paths
// have the expected effect.
// If used, this function must be called before ConfigAutosavePath and
// DefaultStorage are being used.
func UpdateStoragePaths() {
	ConfigAutosavePath = filepath.Join(AppConfigDir(), "autosave.json")
	DefaultStorage = &certmagic.FileStorage{Path: AppDataDir()}
}

// Here everything stays as it is now

// ConfigAutosavePath is the default path to which the last config will be persisted.
var ConfigAutosavePath = filepath.Join(AppConfigDir(), "autosave.json")

// DefaultStorage is Caddy's default storage module.
var DefaultStorage = &certmagic.FileStorage{Path: AppDataDir()}
```

This solution should be fine as the program flow does seem to first refer to this stuff by using `caddy.ConfigAutosavePath`  at `cmd/commandfuncs.go` in line 185 (and only when `--resume` is set. The update of the variables would have already happened in line 170 for the example of the `caddy run` command.

_If you think this issue should be solved and this would be an appropriate solution, I would implement it and open a PR.
I'm open for suggestions for a better or cleaner solution._

**Comment by francislavoie** [comment]
Thanks for opening an issue! We'll look into this.

It's not immediately clear to us what is going on, so we'll need your help to understand it better.

Ideally, we need to be able to reproduce the bug _in the most minimal way possible_. This allows us to write regression tests to verify the fix is working. If we can't reproduce it, then you'll have to test our changes for us until it's fixed -- and then we can't add test cases, either.

I've attached a template below that will help make this easier and faster! It will ask for some information you've already provided; that's OK, just fill it out the best you can. :+1:

I've also included some helpful tips below the template. Feel free to let me know if you have any questions!

Thank you again for your report, we look forward to resolving it!

## Template

<pre><code>## 1. Environment

### 1a. Operating system and version

```
paste here
```


### 1b. Caddy version (run `caddy version` or paste commit SHA)

```
paste here
```


### 1c. Go version (if building Caddy from source; run `go version`)

```
paste here
```


## 2. Description

### 2a. What happens (briefly explain what is wrong)




### 2b. Why it's a bug (if it's not obvious)




### 2c. Log output

```
paste terminal output or logs here
```



### 2d. Workaround(s)




### 2e. Relevant links




## 3. Tutorial (minimal steps to reproduce the bug)




</code></pre>

## Helpful tips

1. **Environment**: Please fill out your OS and Caddy versions, even if you don't think they are relevant. (They are _always_ relevant.) If you built Caddy from source, provide the commit SHA and specify your exact Go version.

2. **Description:** Describe at a high level what the bug is. What happens? Why is it a bug? Not all bugs are obvious, so convince readers that it's actually a bug.
	- **2c) Log output:** Paste terminal output and/or complete logs in a code block. [DO NOT REDACT INFORMATION except for credentials.](https://caddy.community/t/how-to-get-help-with-caddy-more-effectively/5222)
	- **2d) Workaround:** What are you doing to work around the problem in the meantime? This can help others who encounter the same problem, until we implement a fix.
	- **2e) Relevant links:** Please link to any related issues, pull requests, docs, and/or discussion. This can add crucial context to your report.

3. **Tutorial:** What are the _minimum required specific steps_ someone needs to take in order to experience the same bug? Your goal here is to make sure that anyone else can have the same experience with the bug as you do. You are writing a tutorial, so make sure to carry it out yourself before posting it. Please:
	- Start with an empty config. Add _only_ the lines/parameters that are _absolutely required_ to reproduce the bug.
	- Do not run Caddy inside containers.
	- Run Caddy manually in your terminal; do not use systemd or other init systems.
	- If making HTTP requests, avoid web browsers. Use a simpler HTTP client instead, like `curl`.
	- Do not redact any information from your config (except credentials). Domain names are public knowledge and [often necessary for quick resolution of an issue](https://caddy.community/t/how-to-get-help-with-caddy-more-effectively/5222)!
	- Note that ignoring this advice may result in delays, or even in your issue being closed. 😞 Only actionable issues are kept open, and if there is not enough information or clarity to reproduce the bug, then the report is not actionable.

Example of a tutorial:

<blockquote>Create a config file:

```
{ ... }
```

Open terminal and run Caddy:

```
$ caddy ...
```

Make an HTTP request:

```
$ curl ...
```

Notice that the result is ___ but it should be ___.</blockquote>

**Comment by pascal-zarrad** [comment]
@francislavoie I updated the description. Accidently hit return when writing the title which created an empty issue.

PS: Issue templates would be great. Maybe the template you posted could just be added?

**Comment by mholt** [comment]
> PS: Issue templates would be great. Maybe the template you posted could just be added?

It's a saved reply. Most people ignore help/issue templates. We only use it when it's needed after the initial post.

**Comment by pascal-zarrad** [comment]
The mentioned possible solution on a fork [pascal-zarrad/caddy](https://github.com/pascal-zarrad/caddy/commit/813ebf8f41c6308edfd03ab1d181c92e0437dc39) that could be PR'd

**Comment by mholt** [comment]
Thanks for filling out the template.

To clarify, the issue is primarily with the path to the autosave file, is that correct?

The config autosave is intentionally stored to a "hard-coded" path (that doesn't use the configured storage), since autosaved config is done per-instance, not a single config for the whole cluster sharing the same storage (that would get problematic).

**Comment by pascal-zarrad** [comment]
# TL;DR;
With `XDG_DATA_HOME` and `XDG_DATA_HOME` set on a system level or passed when starting Caddy, the locations where the config and data files of Caddy are located are the defined ones.

If you start Caddy and define a `.env` using `--envfile` that should override  those values,  Caddy will either use the system defined ones or fallback to $HOME when the XDG_* values are not defined.

The expected behavior would be that Caddy would use the values from the env file and create it's `caddy` directory for the modified path under the specified path(s).

The requested changes (that you can see on my fork) would allow `--envfile` to change that behavior. The necessary code changes would only affect Caddy when `--envfile` is specified. As there are no command line options to specify the Caddy data and config paths , I think supporting changing the paths in the `.env` would be nice. 

If you don't like overwriting such a crucial value, a command line option would be another possbility. But I think as this would still require an adjustment of the `ConfigAutosavePath` and `DefaultStorage` after they have been initialized.

Conclusion:
_So in the end, the request is the possibility to define the location of the config and data homes without needing to change the users or systems variables or putting them in front of the startup command.  Either the easy addition to the .env file or a CLI argument may should provide this functionality_ 

# Explaination
My target is to choose a specific directory for the files that Caddy uses internally - but using the .env files.

The documentations says the following environment variables are used for [file locations](https://caddyserver.com/docs/conventions#file-locations):
 - HOME = Fallback for XDG_DATA_HOME and XDG_CONFIG_HOME
 - XDG_DATA_HOME = Storage for TLS certs and other important stuff
 - XDG_CONFIG_HOME = Path for where the autosave.json (Path: <any_path>/caddy/autosave.json

Caddyfile for test:
```
localhost

respond "Hello, world!"
```

### Default Behaviour
If I start a basic Caddy with the minimal configuration and default environment variables (no XDG_* set) (on a server or locally) it will have the following behaviour:
 - Creates an `/home/pascal/.local/share/caddy/` directory where everything is stored

### Behaviour with environment variables
Now let's say I have the case that I need the data and autosave config stored in a specific other directory, for demonstration purposes we define `/home/pascal/caddydata` as our directory where everything created by Caddy should land in. The data should lay in a `data` and the `autosave.json` in a `config` subdirectory.

This means we define:
 - XDG_DATA_HOME  = /home/pascal/caddydata/data
 - XDG_CONFIG_HOME  = /home/pascal/caddydata/config

 We start Caddy with: `XDG_DATA_HOME=/home/pascal/caddydata/data XDG_CONFIG_HOME=/home/pascal/caddydata/config ./.caddy/caddy run --config /home/pascal/Caddyfile` which will set the variables.

**The behaviour now:**
 - Creates an `/home/pascal/caddydata/config` directory where the `autosave.json` is stored in a `caddy` subdir (as expected)
 - Creates an `/home/pascal/caddydata/data` directory where the TLS data and other stuff is stored in a `caddy` subdir (as expected)

Note: That `caddy` is always created as subdirectory is fine, its as defined in the docs.

### Behaviour with env file
Also in this case: For demonstration purposes we define `/home/pascal/caddydata` as our directory where everything created by Caddy should land in. The data should lay in a `data` and the `autosave.json` in a `config` subdirectory.
 
We create an env file:
```
XDG_DATA_HOME=home/pascal/caddydata/data
XDG_DATA_HOME=/home/pascal/caddydata/config
# Here could be other variables defined that are required
```

 We start Caddy with: `XDG_DATA_HOME=/home/pascal/caddydata/data XDG_CONFIG_HOME=/home/pascal/caddydata/config ./.caddy/caddy run --envfile /home/pascal/.env --config /home/pascal/Caddyfile` which will set the variables.
The .env location is just an example, wouldn't make sense at all in most cases to store it in /home/<user>/.env.

**The behaviour now:**
We would now expected the same result for where the files are stored as under "Behaviour with environment variables".

Instead we get the following behavior:
 - Creates an `/home/pascal/.local/share/caddy/` directory where everything is stored

**Comment by francislavoie** [comment]
> The mentioned possible solution implemented on a fork [pascal-zarrad/caddy](https://github.com/pascal-zarrad/caddy/commit/813ebf8f41c6308edfd03ab1d181c92e0437dc39) that could be PR'd

Please go ahead and make the PR. That looks good to me!

**Comment by pascal-zarrad** [comment]
@francislavoie PR has been created, rebase is pending and will be done later the day.
